### DISCUSSION TITLE
How to generate a SPEF (`write_spef`) including coordinate information?

### DISCUSSION THREAD

**Discussion by xlindo** [The-OpenROAD-Project/OpenROAD#3390] [Q&A]
As the default flow, there is no coordinate information in `*CONN` following each `PORT` in the generated `6_final.spef`, which is not in coincidence with a normal SPEF format.

E.g., no `*C` bolow,

![image](https://github.com/The-OpenROAD-Project/OpenROAD/assets/8579640/488691da-2217-4dae-980b-da833f9bbf79)

```
...

*CONN
*I *572:B I *D sky130_fd_sc_hd__nand2_1
*I *571:Y O *D sky130_fd_sc_hd__nand2_1

...
```

Compared with another example in a book,

```
*D_NET *5423 2.69358 

*CONN 
*I *14207:D I *C 21.7450 94.3150 
*I *14205:D I *C 21.7450 90.4900 
*I *14211:Q O *C 21.4900 83.8800 *D DFFQX1 

*CAP 
1 *5423:10107 *547:12722 0.202686 
2 *5423:10107 *5116:10594 0.104195

....
```

Any boss help me?

**Comment by maliberty** [comment]
Coordinates are optional in spef so this is valid.  There is code to generate coordinates but we've never used it.  We could try enabling but I can't make any guarantees about what it will produce.  Do you have a use model for these or are you just comparing to what a book shows?

**Reply by xlindo** [reply]
> Coordinates are optional in spef so this is valid. There is code to generate coordinates but we've never used it. We could try enabling but I can't make any guarantees about what it will produce. Do you have a use model for these or are you just comparing to what a book shows?

I would like to generate the coordinates for later usage. I saw the parameter `nodeCoord` is in the function `writeSPEF` (or `writeBlock`) in `netRC.cpp`, but the usage is as 

```cpp
void
write_spef(const char* file,
           const char* nets,
           int net_id)
{
  Ext* ext = getOpenRCX();
  Ext::SpefOptions opts;
  opts.file = file;
  opts.nets = nets;
  opts.net_id = net_id;
  
  ext->write_spef(opts);
}
```

in `ext.i`, where I guess if the coordinate information is never parsed? So I am not sure how to enable and if it really works.

**Reply by xlindo** [reply]
> Coordinates are optional in spef so this is valid. There is code to generate coordinates but we've never used it. We could try enabling but I can't make any guarantees about what it will produce. Do you have a use model for these or are you just comparing to what a book shows?

FYI, the example with coordinates is at page 545 of _Static Timing Analysis for Nanometer Designs: A Practical Approach. Boston, MA: Springer US, 2009. doi: [10.1007/978-0-387-93820-2](https://doi.org/10.1007/978-0-387-93820-2)._

Also, the coordinates are important for RC estimation as the training resources.

**Comment by maliberty** [comment]
I've added a -coordinates option to write_spef in https://github.com/The-OpenROAD-Project/OpenROAD/pull/3434
